### PR TITLE
Support for load replay gain info from CUE

### PR DESCRIPTION
**Pull request by PSyton**
ReplayGain refactoring. (Create class for store and work with RG data)
CCueDocument refactoring.
Support for load replay gain info from CUE

### CODE REVIEW

**Comment by notspiff** [comment]
if the summary contains 3 separate logical changes, there should be 3 separate commits.

**Comment by notspiff** [comment]
thanks :)

**Comment by FernetMenta** [comment]
do you have a chance to build on Linux? I don't see any makefile updated for the new files.

File: xbmc/music/CueInfoLoader.h [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#pragma once
+/*
+ *      Copyright (C) 2005-2013 Team XBMC
```
Comment (FernetMenta) [inline xbmc/music/CueInfoLoader.h]: @MartijnKaijser should this be 2015?

Comment (MartijnKaijser) [inline xbmc/music/CueInfoLoader.h]: Might as well. We don't have a real reason for it after some online searching as we have all code online with date of contributions

**Comment by PSyton** [comment]
No linux machine at this time. Only win and OS X.

**Comment by FernetMenta** [comment]
please squash. build needs to work with any given sha checked out or you get into trouble when bisecting

File: xbmc/music/CueInfoLoader.cpp [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ */
+#include "CueInfoLoader.h"
+
+void CueInfoLoader::Load(const std::string& aStrCuesheet, CFileItemPtr aFileItem)
+{
+  if (aStrCuesheet.empty())
+    return;
+  
+  std::string p = aFileItem->GetPath();
```
Comment (Paxxi) [inline xbmc/music/CueInfoLoader.cpp]: Is this some leftover? I don't see it used anywhere

File: xbmc/music/tags/ReplayGain.h [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+class ReplayGain
+{
+public:
+  enum Type {
+    NONE = 0,
+    ALBUM,
+    TRACK
+  };
+public:
```
Comment (Paxxi) [inline xbmc/music/tags/ReplayGain.h]: this can be removed as you already have one further up

File: xbmc/music/CueInfoLoader.cpp (hunk 2 above)
Comment (PSyton) [inline xbmc/music/CueInfoLoader.cpp]: Yes. Forgot to remove this.

**Comment by FernetMenta** [comment]
could you please rebase and resolve the conflict. Update of makefile and project files should not be done in a separate commit because HEAD~1 of this branch won't build.

**Comment by PSyton** [comment]
Some files should be added to 9055e6e some to e8f5c1d because HEAD~2 and HEAD~3 won't build too. @notspiff ask to split this code into 3 commits. Now I need squash it again? Why so difficult?

**Comment by FernetMenta** [comment]
no single commit should break build. as I mentioned before bisecting is hard if build fails. a single change set should be self-contained and neither break build nor runtime

**Comment by PSyton** [comment]
Cherry-picked project updates

**Comment by FernetMenta** [comment]
the VS project is broken on this branch. does not show any filters

**Comment by PSyton** [comment]
Project fixed.

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by PSyton** [comment]
Added stdlib.h for atof. Non windows builds should be fixed.

**Comment by mkortstiege** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
why did you force push again?

**Comment by PSyton** [comment]
Because when I've fixed 'atof', Makefile in xbmc/music has been occasionally reverted.

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by PSyton** [comment]
Fixed XCode project for ATV2

**Comment by PSyton** [comment]
XCode project fix

**Comment by FernetMenta** [comment]
jenkins build this please

File: xbmc/cores/paplayer/AudioDecoder.cpp [hunk 4]
```diff
@@ -102,9 +102,30 @@ bool CAudioDecoder::Create(const CFileItem &file, int64_t seekOffset)
...
-  if (file.HasMusicInfoTag() && file.GetMusicInfoTag()->GetDuration())
-    m_codec->SetTotalTime(file.GetMusicInfoTag()->GetDuration());
+  if (file.HasMusicInfoTag())
+  {
+    // set total time from the given tag
+    if (file.GetMusicInfoTag()->GetDuration())
+      m_codec->SetTotalTime(file.GetMusicInfoTag()->GetDuration());
+
+    // update ReplayGain from the given tag if it's better then original (cuesheet)
+    ReplayGain rgInfo = m_codec->m_tag.GetReplayGain();
```
Comment (FernetMenta) [inline xbmc/cores/paplayer/AudioDecoder.cpp]: 1) a few lines down you made this a reference. why not here?

Comment (FernetMenta) [inline xbmc/cores/paplayer/AudioDecoder.cpp]: include of ReplayGain.h missing

Comment (PSyton) [inline xbmc/cores/paplayer/AudioDecoder.cpp]: 1. We can't use const reference here. This is not read only.
2. MusicInfoTag.h contains include ReplayGain.h and we can't use predeclaration.

Comment (FernetMenta) [inline xbmc/cores/paplayer/AudioDecoder.cpp]: you use ReplayGain here, hence you should include the header.

**Comment by PSyton** [comment]
Added include. But this not helps with ATV2 build

**Comment by FernetMenta** [comment]
@Memphiz do you have an idea why ATV2 build fails?

**Comment by Memphiz** [comment]
Are you guys confusing atv2 with ios? Jenkins doesn't build atv2 anymore - need to check Once i am home again sunday evening...

**Comment by FernetMenta** [comment]
I was referring to IOS-ATV2

**Comment by Memphiz** [comment]
Ok - i got ill and am not Able to Drive back home for a while - as i don't have my hackbook with me i can't look into it for at least a week - stupid flu *narf*

**Comment by PSyton** [comment]
Any news?

**Comment by FernetMenta** [comment]
@PSyton can you rebase. maybe @Memphiz has time to have a look at iOS

**Comment by Memphiz** [comment]
xcode sync was bad ... @mkortstiege somehow you ReplayGain.cpp was only added to osx and atv2 targets but not to ios target. Also CueInfoLoader.cpp was not sorted well in the list but was above the subdirs ... psyton please cherry-pick https://github.com/Memphiz/xbmc/commit/e6f3c48101c290aedbc336bb354e02c374562f1c and squash it into your commits. After a rebase it should work then (expect a conflict in the windows project filters which is easy to solve i think...)

**Comment by PSyton** [comment]
Done

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by mkortstiege** [comment]
Build error is unrelated. @Memphiz sorry. Looks like I totally fucked this one up then.

**Comment by Memphiz** [comment]
no problem - i really appreciate your syncing work ;)

**Comment by FernetMenta** [comment]
@Memphiz thanks! btw: your work on Jenkins is awesome

**Comment by PSyton** [comment]
Sorry http://jsfiddle.net/5m24e927/1/

**Pull request merged**